internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,297 @@
+	sl.mlMux.Lock()
+	sl.mlSpeakersIPs = iplist
+	sl.mlMux.Unlock()
+
+	// Update mlSpeakersIPs in the background.
+	go sl.speakersIPsLoop()
+
+	// Every one minute, try to rejoin.
+	// This will join nodes that leave the cluster only for a few seconds every 1
+	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
+	// take a while longer.
+	ticker := time.NewTicker(1 * time.Minute)
+
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.mlJoin()
+		case <-sl.mlJoinCh:
+			sl.mlJoin()
+		}
+	}
+}
+
+func (sl *SpeakerList) members() map[string]struct{} {
+	members := make(map[string]struct{})
+
+	for _, node := range sl.mList.Members() {
+		ip := node.Addr.String()
+		members[ip] = struct{}{}
+	}
+
+	return members
+}
+
+// mlJoin joins this memberlist cluster with other speakers pods that are not
+// currently member of this cluster. In other words, it does a memberlist.Join()
+// with the IPs in mlSpeakersIPs that are not members of the cluster.
+func (sl *SpeakerList) mlJoin() {
+	// Calculate IPs to join, only IPs not currently members of the memberlist cluster.
+	var joinIPs []string
+
+	members := sl.members()
+
+	sl.mlMux.Lock()
+	defer sl.mlMux.Unlock()
+
+	for _, ip := range sl.mlSpeakersIPs {
+		// If an ip is not member of the cluster, add it to joinIPs.
+		if _, isMember := members[ip]; !isMember {
+			joinIPs = append(joinIPs, ip)
+		}
+	}
+
+	if len(joinIPs) == 0 {
+		// Not logging here to not spam the logs.
+		return
+	}
+
+	nr, err := sl.mList.Join(joinIPs)
+	if err != nil || nr != len(joinIPs) {
+		sl.l.Log("op", "memberDiscovery", "msg", "partial join", "joined", nr, "expected", len(joinIPs), "error", err)
+	} else {
+		sl.l.Log("op", "Member detection", "msg", "memberlist join succesfully", "number of other nodes", nr)
+	}
+}
+
+// Rejoin initiates a discovery and joins all the speakers to the memberlist
+// cluster.
+func (sl *SpeakerList) Rejoin() {
+	if sl.mList == nil {
+		return
+	}
+
+	select {
+	case sl.mlJoinCh <- struct{}{}:
+		sl.l.Log("op", "memberDiscovery", "msg", "triggering discovery")
+	default:
+		sl.l.Log("op", "memberDiscovery", "msg", "previous discovery in progress - doing nothing")
+	}
+
+	return
+}
+
+// UsableSpeakers returns a map of usable speaker nodes.
+func (sl *SpeakerList) UsableSpeakers() map[string]bool {
+	if sl.mList == nil {
+		return nil
+	}
+	activeNodes := map[string]bool{}
+	for _, n := range sl.mList.Members() {
+		activeNodes[n.Name] = true
+	}
+	return activeNodes
+}
+
+// Stop terminates an already initialized speakerlist.
+func (sl *SpeakerList) Stop() {
+	if sl.mList == nil {
+		return
+	}
+
+	sl.l.Log("op", "shutdown", "msg", "leaving memberlist cluster")
+	err := sl.mList.Leave(time.Second)
+	sl.l.Log("op", "shutdown", "msg", "left memberlist cluster", "error", err)
+	err = sl.mList.Shutdown()
+	sl.l.Log("op", "shutdown", "msg", "memberlist shutdown", "error", err)
+}
+
+func event2String(e memberlist.NodeEventType) string {
```

Comment:
just to be clear: this is what is already merged in main branch.

Converts `memberlist.NodeEventType` in a string to print the event as a string (the struct is not printed nicely otherwise, IIRC).

This function is used to print a log line (string) when a node joins or leaves.

Reply:
@johananl re-ping?